docs/job-traces.md

```diff
@@ -58,6 +58,50 @@ pipeline {
 }
 ````
 
+### Custom spans
+
+Custom spans can be defined using the `withNewSpan` step, which accepts the following parameters
+* `label`
+  * the label of the span
+  * the value is a `string`
+* `attributes`
+  * a list of attributes, defined the same way as in the `withSpanAttributes` step
+  * ```groovy
+    attributes: ([
+        spanAttribute(key: 'modules', value: '2'),
+        spanAttribute(key: 'command', value: 'mvn clean install')
+    ])
+    ```
+* `setAttributesOnlyOnParent`
+  * flag used to define whether to inherit the provided attributes to the children spans or not
+  * `true` by default, all user-defined attributes for a span are passed on to children spans
+  * the value is a boolean, `true` or `false`
+
+Example definitions:
+
+* All parameters provided
+    ```groovy
+    stage('build') {
+        withNewSpan(label: 'custom-build-span', attributes: ([
+            spanAttribute(key: 'modules', value: '2'),
+            spanAttribute(key: 'command', value: 'mvn clean install')
```

Comment:
Out of curiosity, could this be less verbose as a key and value instead of a spanAttribute object?

Something like the below pseudo-code:
```
 withNewSpan(label: 'custom-build-span', attributes: ([
            'modules': '2',
            'command': 'mvn clean install'])
```

The declarative syntax might benefit from a shorter syntax. If users decide to use a bunch of attributes, with the current proposal, there will be 26 chars extra per entry.

What do you think?

Reply:
Hi @v1v, thanks for the review!

> could this be less verbose as a key and value instead of a spanAttribute object?

Not at the moment, but it's doable.

In my current approach, I'm reusing the code for the `SpanAttributeStepExecution` and it works the same way as in `WithSpanAttributesStep` which has this constructor

```java
    @DataBoundConstructor
    public WithSpanAttributesStep(List<SpanAttribute> spanAttributes) {
        this.spanAttributes = spanAttributes;
    }
``` 

for that reason, the `WithNewSpanStep` also accepts a `List<SpanAttribute>`. 

Because it's annotated with `@DataBoundConstructor`, you can't overload it but if the parameter type changes to `Object attributes`, it can also accept a map and that would achieve the behavior that you want. We can convert the map to a `List<SpanAttribute>` and keep reusing the existing code.

The question is, do we want to make the change only in `WithNewSpanStep` or should we make it in `WithSpanAttributesStep` to make the change more global and keep the behavior consistent?

Reply:
I did some testing and it turns out that it's not that straight forward. Jenkins doesn't like the type-cast from `Object` to a `SpanAttribute` element of a List. No issues with using a Map, we can get the values and then create a `List<SpanAttribute>` but that will break consistency with the other steps.

Reply:
Hi @v1v, any thoughts on this?

Reply:
Thanks Christos for your analysis. It's totally fine